typescript/cli/src/context/altvm.ts

```diff
@@ -0,0 +1,202 @@
+      );
+    }
+
+    const { provider } = ALT_VM_SUPPORTED_PROTOCOLS[metadata.protocol]!;
+    return provider.connect(metadata.rpcUrls[0].http);
+  }
+}
+
+export class AltVMSignerFactory implements AltVM.ISignerFactory {
+  private readonly metadataManager: ChainMetadataManager;
+  private readonly chains: ChainMap<AltVM.ISigner>;
+
+  private constructor(
+    metadataManager: ChainMetadataManager,
+    chains: ChainMap<AltVM.ISigner>,
+  ) {
+    this.metadataManager = metadataManager;
+    this.chains = chains;
+  }
+
+  public getSupportedProtocols(): ProtocolType[] {
+    return Object.keys(ALT_VM_SUPPORTED_PROTOCOLS) as ProtocolType[];
+  }
+
+  public supports(protocol: ProtocolType) {
+    return !!ALT_VM_SUPPORTED_PROTOCOLS[protocol];
+  }
+
+  public getGas(protocol: ProtocolType) {
+    if (!this.supports(protocol)) {
+      throw new Error(`Protocol type ${protocol} not supported in AltVM`);
+    }
+
+    const { gas } = ALT_VM_SUPPORTED_PROTOCOLS[protocol]!;
+    return gas;
+  }
+
+  public get(chain: string): AltVM.ISigner {
+    const protocol = this.metadataManager.getProtocol(chain);
+
+    if (!this.supports(protocol)) {
+      throw new Error(
+        `Chain ${chain} with protocol type ${protocol} not supported in AltVM`,
+      );
+    }
+
+    if (!this.chains[chain]) {
+      throw new Error(`AltVM was not initialized with chain ${chain}`);
+    }
+
+    return this.chains[chain];
+  }
+
+  private static async loadPrivateKey(
+    key: SignerKeyProtocolMap,
+    strategyConfig: Partial<ExtendedChainSubmissionStrategy>,
+    protocol: ProtocolType,
+    chain: string,
+  ): Promise<string> {
+    // 1. First try to get private key from --key.{protocol} flag
+    if (key[protocol]) {
+      return key[protocol]!;
+    }
+
+    // 2. If no key flag was provided we check if a strategy config
+    // was provided for our chain where we can read our private key
+    if (strategyConfig[chain]) {
+      const rawConfig = strategyConfig[chain]!.submitter;
+      if (!isJsonRpcSubmitterConfig(rawConfig)) {
+        throw new Error(
+          `found unknown submitter in strategy config for chain ${chain}`,
+        );
+      }
+
+      if (!rawConfig.privateKey) {
+        throw new Error(
+          `missing private key in strategy config for chain ${chain}`,
+        );
+      }
+
+      return rawConfig.privateKey;
+    }
+
+    // 3. Finally, if no key flag or strategy was provided we prompt the user
+    // for the private key
+    return password({
+      message: `Please enter the private key for chain ${chain} (will be re-used for other chains with the same protocol type)`,
+    });
+  }
+
+  public static async createSigners(
+    metadataManager: ChainMetadataManager,
+    chains: string[],
+    key: SignerKeyProtocolMap,
+    strategyConfig: Partial<ExtendedChainSubmissionStrategy>,
+  ) {
+    const signers: ChainMap<AltVM.ISigner> = {};
+
+    for (const chain of chains) {
```

Comment:
Why do we create signers up-front, while creating providers on demand? I mean that `providerFactory.get()` instantiates and connects a provider but `signerFactory.get()` just returns a signer that was created at some initialisation time.

Reply:
Yes that is exactly the difference between the provider and signer factory. The difference exists because for all signing commands (core/warp deploy,apply) we know exactly for which chains we need signers for. On the provider side it is expected to create providers dynamically if requested. This structure was introduced by the original `multiProvider` which gives you providers dynamically but initializes signers for signing commands when the chains are known. Else of course our AltVM factory classes would be much smaller and more generic